domain/src/main/java/com/teamagam/gimelgimel/domain/messages/DisplayMessagesInteractor.java

```diff
@@ -50,41 +50,41 @@
                 mMessagesRepository.getMessagesObservable(),
                 this::showMessage);
 
-        DisplaySubscriptionRequest displayShownStatus = factory.create(
-                mDisplayedEntitiesRepository.getObservable(),
-                this::updateRelatedMessageDisplayStatus);
+        DisplaySubscriptionRequest updateMapDisplayStatusChanges = factory.create(
+                getMessageMapDisplayChanges(),
+                this::showMessage);
 
-        return Arrays.asList(displayMessages, displayShownStatus);
+        return Arrays.asList(displayMessages, updateMapDisplayStatusChanges);
     }
 
     private void showMessage(Message message) {
-        mDisplayer.show(message, mMessagesUtil.isMessageFromSelf(message));
+        mDisplayer.show(createMessagePresentation(message));
     }
 
-    private void updateRelatedMessageDisplayStatus(GeoEntityNotification geoEntityNotification) {
-        mMapper.getMessageId(geoEntityNotification.getGeoEntity().getId())
+    private Observable<Message> getMessageMapDisplayChanges() {
+        return mDisplayedEntitiesRepository.getObservable()
+                .map(geoEntityNotification -> geoEntityNotification.getGeoEntity().getId())
+                .flatMap(mMapper::getMessageId)
                 .flatMap(mMessagesRepository::getMessage)
-                .filter(message -> message != null)
-                .subscribe(message -> {
-                    updateMessageDisplayStatus(geoEntityNotification, message);
-                });
+                .filter(m -> m != null);
     }
 
-    private void updateMessageDisplayStatus(GeoEntityNotification geoEntityNotification,
-                                            Message message) {
-        if (geoEntityNotification.getAction() == GeoEntityNotification.ADD) {
-            mDisplayer.messageShownOnMap(message);
-        } else if (geoEntityNotification.getAction() == GeoEntityNotification.REMOVE) {
-            mDisplayer.messageHiddenFromMap(message);
+    private MessagePresentation createMessagePresentation(Message message) {
+        boolean isShownOnMap = isShownOnMap(message);
+        boolean isFromSelf = mMessagesUtil.isMessageFromSelf(message);
+        return new MessagePresentation(message, isFromSelf, isShownOnMap);
+    }
+
+    private boolean isShownOnMap(Message message) {
+        if (!(message instanceof BaseMessageGeo)) {
+            return false;
         }
+        BaseMessageGeo bmg = (BaseMessageGeo) message;
+        return !mDisplayedEntitiesRepository.isNotShown(bmg.getGeoEntity());
```

Comment:
Why doesn't DisplayedEntitiesRepository have a method "isEntityShown"? its implementation has it but doesn't expose it because of the interface. So you find yourself negating twice...